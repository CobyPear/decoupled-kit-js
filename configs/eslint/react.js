/** @type {import('eslint').ESLint.ConfigData} */
module.exports = {
	plugins: ['react'],
	extends: [
		'./index.js',
		'plugin:react/recommended',
		'plugin:react-hooks/recommended',
		'plugin:react/jsx-runtime',
	],
	settings: {
		react: {
			version: 'detect',
		},
	},
};
