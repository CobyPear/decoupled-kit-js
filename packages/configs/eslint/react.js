module.exports = {
	plugins: ['react'],
	extends: [
		'./index.js',
		'plugin:react/recommended',
		'plugin:react-hooks/recommended',
	],
	settings: {
		react: {
			version: 'detect',
		},
	},
};
