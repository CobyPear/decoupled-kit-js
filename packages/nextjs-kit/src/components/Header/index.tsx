import {
	Header as ReactKitHeader,
	isNavItemArray,
	type NavHeaderProps,
	type NavItem,
} from '@pantheon-systems/react-kit/components/Header';
import Link from 'next/link';

const Header = (props: NavHeaderProps) => {
	return (
		<ReactKitHeader
			{...props}
			linkComponent={props?.linkComponent ? props.linkComponent : Link}
		/>
	);
};

export { Header, NavHeaderProps, NavItem, isNavItemArray };
