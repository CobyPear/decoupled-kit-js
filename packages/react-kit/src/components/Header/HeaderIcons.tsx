export const CloseSVG = () => {
	return (
		<svg
			width="24"
			height="24"
			viewBox="0 0 40 40"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
		>
			<circle cx="20" cy="20" r="20" fill="none" />
			<path d="M11 11L29 29" stroke="black" strokeWidth="2" />
			<path d="M29 11L11 29" stroke="black" strokeWidth="2" />
		</svg>
	);
};

export const HamburgerMenuSVG = () => {
	return (
		<svg
			aria-label="open"
			width="24"
			height="24"
			viewBox="0 0 40 40"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
		>
			<circle cx="20" cy="20" r="20" fill="none" />
			<path
				d="M9 12H31"
				stroke="#171717"
				strokeWidth="2"
				strokeLinecap="round"
			/>
			<path
				d="M9 20H31"
				stroke="#171717"
				strokeWidth="2"
				strokeLinecap="round"
			/>
			<path
				d="M9 28L31 28"
				stroke="#171717"
				strokeWidth="2"
				strokeLinecap="round"
			/>
		</svg>
	);
};
