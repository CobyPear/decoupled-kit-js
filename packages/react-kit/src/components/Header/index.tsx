export { Header } from './Header';
export * from './props';
