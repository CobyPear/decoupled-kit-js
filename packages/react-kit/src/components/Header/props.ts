import { Props as FocusTrapProps } from 'focus-trap-react';

/**
 * A navigation item is a tuple of a label and a href.
 * @example
 * ```ts
 * const navItems = [['Home', '/', 'About', '/about',]]
 * ```
 */
export type NavItem = [string, string];

export type NavHeaderProps = Readonly<{
	/**
	 * A logo component to render in the header.
	 * @remarks should be wrapped in a link to the homepage or other relevant page.
	 * @example
	 * ```tsx
	 * export const Logo = ({ exampleLogo }: { exampleLogo: string }) => {
	 * 	// use the link and image component from your framework if applicable.
	 * 	return (
	 * 		<a href="/">
	 * 			<img src={exampleLogo} width="48" height="48" alt="Example Logo" />
	 * 		</a>
	 * 	);
	 * };
	 * ```
	 *
	 */
	Logo: React.ElementType;
	/**
	 * Items to render for the main navigation. Can be a tuple of label and href or a ReactNode.
	 */
	mainNavItems?: NavItem[] | React.ReactNode;
	/**
	 * Secondary navigation menu. Can be a tuple of label and href or a ReactNode
	 */
	secondaryNavItems?: NavItem[] | React.ReactNode;
	/**
	 * Styles to be applied to the mobile nav overlay.
	 */
	overlayStyles?: string;
	/**
	 * Styles to be applied to the omnipresent navbar.
	 */
	navbarStyles?: string;
	/**
	 * Options to pass to the FocusTrap component used to trap focus within the mobile nav overlay.
	 */
	focusTrapOptions?: FocusTrapProps['focusTrapOptions'];
}>;

/**
 * Type guard for NavItem
 * @param arg the argument to check
 * @returns true if the arg is a {@link NavItem}
 */
export const isNavTuple = (arg: unknown): arg is NavItem => {
	if (arg instanceof Array) {
		return arg
			.map((item) => {
				return (
					item instanceof Array &&
					item.length === 2 &&
					typeof item[0] === 'string' &&
					typeof item[1] === 'string'
				);
			})
			.reduce((acc, curr) => acc && curr, true);
	}
	return false;
};
