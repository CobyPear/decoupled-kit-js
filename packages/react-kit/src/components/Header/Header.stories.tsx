import PantheonLogo from '@assets/pantheon-fist-blk.png';
import { Button } from '@components/Button';
import { Header as NavHeader } from '@components/Header';
import { Row } from '@components/Row';
import type { Meta, StoryObj } from '@storybook/react';
import clsx from 'clsx';

const Logo = () => {
	return (
		<>
			<a href="/">
				<img
					src={PantheonLogo}
					width="40"
					height="40"
					alt="Pantheon Systems"
					className="lg:rk-mr-8"
				/>
			</a>
		</>
	);
};

const Header = ({
	mainNavItems,
	secondaryNavItems,
	Logo,
}: {
	mainNavItems: React.ReactNode;
	secondaryNavItems: React.ReactNode;
	Logo: React.ElementType;
	manyItems?: boolean;
}) => {
	return (
		<Row type="flex" className="rk-mx-auto rk-bg-white">
			<NavHeader
				overlayStyles="rk-bg-white"
				Logo={Logo}
				mainNavItems={mainNavItems}
				secondaryNavItems={secondaryNavItems}
			/>
		</Row>
	);
};

const MainNavItems = ({ manyItems }: { manyItems?: boolean }) => {
	const navItems = [
		['Home', '/'],
		['Articles', '/articles'],
		['Pages', '/pages'],
	];
	if (manyItems) {
		navItems.push(
			['About', '/about'],
			['Services', '/services'],
			['Contact', '/contact'],
			['Blog', '/blog'],
			['Portfolio', '/portfolio'],
			['FAQ', '/faq'],
			['Shop', '/shop'],
			['Events', '/events'],
			['Gallery', '/gallery'],
			['Team', '/team'],
			['Testimonials', '/testimonials'],
			['Career', '/career'],
			['Terms of Service', '/terms'],
			['Privacy Policy', '/privacy'],
			['Support', '/support'],
			['Login', '/login'],
			['Register', '/register'],
			['Dashboard', '/dashboard'],
			['Logout', '/logout'],
		);
	}
	const Items = navItems.map(([label, href]) => (
		<li
			className="rk-mx-2 rk-mb-8 rk-w-full rk-justify-start rk-text-lg rk-text-black lg:rk-mb-0 lg:rk-w-fit"
			key={label}
		>
			<a
				className={clsx(
					'rk-link-hover rk-w-full rk-text-left',
					'rk-flex rk-w-full rk-min-w-full sm:rk-w-fit',
					href === '/' && 'rk-font-bold',
				)}
				href={href}
			>
				{label}
			</a>
		</li>
	));

	return Items;
};

const SecondaryNavItems = () => {
	return (
		<>
			<li>
				<hr
					className="divide-x rk-my-8 rk-ml-auto rk-flex rk-w-full rk-min-w-full lg:rk-hidden"
					key="hr"
				/>
			</li>
			<li
				className="rk-mb-3 rk-mr-auto lg:rk-mb-0 lg:rk-ml-auto lg:rk-mr-0"
				key="docs"
			>
				<Button
					Element="a"
					href="https://decoupledkit.pantheon.io"
					type="secondary"
				>
					Docs
				</Button>
			</li>
			<li className="rk-mr-auto lg:rk-mx-3 lg:rk-mr-0" key="examples">
				<Button Element="a" href="/examples">
					Examples
				</Button>
			</li>
		</>
	);
};

const meta: Meta<typeof Header> = {
	title: 'Header/Header',
	component: Header,
	parameters: {
		layout: 'fullscreen',
	},
	args: {
		Logo,
		mainNavItems: <MainNavItems />,
		secondaryNavItems: <SecondaryNavItems />,
	},
	argTypes: {
		mainNavItems: {
			control: {
				type: 'radio',
			},
			options: ['Default', 'Many'],
			mapping: {
				Default: <MainNavItems />,
				Many: <MainNavItems manyItems />,
			},
		},
		secondaryNavItems: {
			table: {
				disable: true,
			},
		},
	},
	tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};
