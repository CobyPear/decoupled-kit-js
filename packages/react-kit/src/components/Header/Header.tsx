import { IconButton } from '@components/Button';
import clsx from 'clsx';
import FocusTrap from 'focus-trap-react';
import React, { useEffect, useState } from 'react';
import { CloseSVG, HamburgerMenuSVG } from './HeaderIcons';
import { NavHeaderProps } from './props';

export const Header = ({
	Logo,
	children,
	overlayStyles,
	navbarStyles,
	focusTrapOptions,
}: NavHeaderProps) => {
	const [isOpen, setIsOpen] = useState(false);
	const [windowWidth, setWindowWidth] = useState(window.innerWidth);
	const handleOpen = () => setIsOpen((prev) => !prev);

	const MAX_MOBILE_NAV_WIDTH = 1023;

	useEffect(() => {
		const handleKeyUp = (e: KeyboardEvent) => {
			if (e.key === 'Escape' || e.key === 'Esc') {
				setIsOpen(false);
			}
		};
		const handleResize = () => {
			setWindowWidth(window.innerWidth);
			windowWidth > MAX_MOBILE_NAV_WIDTH && setIsOpen(false);
		};
		setWindowWidth(window.innerWidth);

		window.addEventListener('resize', handleResize);
		window.addEventListener('keyup', handleKeyUp);

		return () => {
			window.removeEventListener('resize', handleResize);
			window.removeEventListener('keyup', handleKeyUp);
		};
	}, [windowWidth]);

	/**
	 * Need to use a ref here so the FocusTrap can work properly.
	 */
	const NavList = React.forwardRef<
		React.ElementRef<'nav'>,
		{ children: React.ReactNode }
	>(({ children }, ref) => {
		return (
			<nav
				ref={ref}
				aria-label="Main"
				className={clsx(navbarStyles, 'rk-z-10 rk-flex rk-flex-row rk-py-4')}
			>
				<h1>
					<Logo />
				</h1>
				<ul
					className={clsx(
						overlayStyles,
						'rk-lg:menu-horizontal rk-w-full rk-items-center',
						isOpen && windowWidth <= MAX_MOBILE_NAV_WIDTH
							? ' rk-menu-vertical rk-absolute rk-left-0 rk-top-[72px] rk-z-10 rk-h-[calc(100%_-_72px)] rk-w-full rk-overflow-y-auto rk-overflow-x-hidden rk-px-4 rk-pt-8 sm:rk-px-6 lg:rk-px-12'
							: 'rk-hidden rk-flex-wrap lg:rk-flex',
					)}
				>
					{children}
				</ul>
				{isOpen && windowWidth <= MAX_MOBILE_NAV_WIDTH ? (
					<IconButton
						aria-hidden={isOpen && windowWidth >= MAX_MOBILE_NAV_WIDTH}
						onClick={handleOpen}
						className="rk-ml-auto lg:rk-flex"
						data-testid="close-nav"
					>
						<CloseSVG />
					</IconButton>
				) : (
					<IconButton
						aria-hidden={!isOpen && windowWidth >= MAX_MOBILE_NAV_WIDTH}
						onClick={handleOpen}
						className="rk-ml-auto rk-flex lg:rk-hidden"
						data-testid="open-nav"
					>
						<HamburgerMenuSVG />
					</IconButton>
				)}
			</nav>
		);
	});

	NavList.displayName = 'NavList';

	return (
		<>
			{windowWidth > MAX_MOBILE_NAV_WIDTH ? (
				<NavList>{children}</NavList>
			) : (
				<FocusTrap
					active={isOpen}
					paused={!isOpen}
					focusTrapOptions={focusTrapOptions}
				>
					<div>
						<NavList>{children}</NavList>
					</div>
				</FocusTrap>
			)}
		</>
	);
};
