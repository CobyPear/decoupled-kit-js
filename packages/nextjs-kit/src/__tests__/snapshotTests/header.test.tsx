import { render } from '@testing-library/react';

import Header from '../../components/header';

/**
 * @vitest-environment jsdom
 */

const headerData = [
  {
    title: 'Home',
    href: '/',
  },
  {
    title: 'Posts',
    href: '/posts',
  },
  {
    title: 'Pages',
    href: '/pages',
  },
];

describe('<Header />', () => {
  it("should render 'header'", () => {
    const { asFragment } = render(<Header headers={headerData} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
