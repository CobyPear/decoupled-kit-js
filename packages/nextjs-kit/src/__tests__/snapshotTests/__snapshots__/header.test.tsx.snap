// Vitest Snapshot v1

exports[`<Header /> > should render 'header' 1`] = `
<DocumentFragment>
  <div
    class="my-0 pt-10 px-5 text-xl"
  >
    <nav>
      <ul
        class="flex flex-row flex-wrap sm:flex-nowrap list-none justify-between max-w-screen-sm mx-auto"
      >
        <li
          class="mr-auto"
        >
          <a
            class="hover:underline"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="mx-4"
        >
          <a
            class="hover:underline"
            href="/posts"
          >
            Posts
          </a>
        </li>
        <li
          class="mx-4"
        >
          <a
            class="hover:underline"
            href="/pages"
          >
            Pages
          </a>
        </li>
      </ul>
    </nav>
  </div>
</DocumentFragment>
`;
